src/TBuffer.cpp

```diff
@@ -2672,49 +2649,127 @@ int TBuffer::wrapLine(int startLine, int screenWidth, int indentSize, TChar& for
                         queue.push(newLine);
                         tempList.append(lineText);
                     }
-                    goto OPT_OUT_CLEAN;
+                    timeList.append(time);
+                    promptList.append(isPrompt);
+                    newLine.clear();
+                    lineText = "";
+                    continue;
                 }
-                newLine.push_back(buffer[line][characterPosition]);
-                lineText.append(lineBuffer[line].at(characterPosition));
-                characterPosition++;
-            }
-            queue.push(newLine);
-            tempList.append(lineText);
+                newLine.push_back(buffer.at(i).at(i2));
+                lineText.append(lineBuffer.at(i).at(i2));
+                // Get current length
+                lineWidth = QFontMetrics(hostFont).horizontalAdvance(lineText);
+                hasContent = true;
+                i2++;
+                if (lineWidth > screenWidth) { // Need to wrap
+                    QTextBoundaryFinder wordFinder(QTextBoundaryFinder::Line, lineText);
+                    wordFinder.setPosition(lineText.size() - 1);
+                    int chopCounter = 1;
+                    if (wordFinder.isAtBoundary() && (wordFinder.boundaryReasons() & QTextBoundaryFinder::BreakOpportunity) != 0) {
+                        wrapPos = lineText.size() - 1;
+                    } else {
+                        while (!(wordFinder.isAtBoundary() && (wordFinder.boundaryReasons() & QTextBoundaryFinder::BreakOpportunity) != 0)) {
+                            wrapPos =  wordFinder.toPreviousBoundary();
+                        }
+                    }
 
-        OPT_OUT_CLEAN:
-            newLine.clear();
-            lineText.clear();
-            indent = 0;
+                    if (wrapPos <= 0) {
+                        chopCounter = 1;
+                    } else {
+                        chopCounter = lineText.size() - wrapPos;
+                    }
+                    
+                    // We chop off however many character we are supposed to
+                    for (int i3 = 0; i3 < chopCounter; ++i3) {
+                        newLine.pop_back();
+                        lineText.chop(1);
+                        i2--;
+                    }
+
+                    queue.push(newLine);
+                    tempList.append(lineText);
+                    timeList.append(time);
+                    promptList.append(isPrompt);
+                    newLine.clear();
+                    lineText = "";
```

Comment:
```suggestion
                    lineText = QString();
```

Qt-recommended way of zeroing out a string